### DISCUSSION TITLE
N64 Controller pak

### DISCUSSION THREAD

**Discussion by Mythris29** [sanni/cartreader#472] [Need help]
I'm having issues writing to an N64 controller memory pak.

I get a consistent CRC (and a different CRC on a different memory pak when I try to save it) so I don't think it's reading incorrectly. However when I try to write the .mpk to another memory pak I get this:

Writing...
//N64/MPK/48.mpk
Done
Verifying...
OK

And then when I try to read the data on an N64 (Mark Kart 64 ghost data in this case) it's not there.

Very confused and I'm hoping somebody can help me with this. I'm on HW5 9.4 firmware.

**Comment by sanni** [comment]
Can you attach the mpk file here? For which version of MK64 is it, US, JAP, PAL?

**Comment by Mythris29** [comment]
Absolutely, here it is. Thank you for looking at this.

[48.zip](https://github.com/sanni/cartreader/files/9354246/48.zip)

**Comment by sanni** [comment]
The file is empty, that's why nothing is there when writing it.

![7171771717171](https://user-images.githubusercontent.com/19169494/184977223-dbc95b46-0402-4202-8948-7b46324e7153.jpg)

**Reply by Mythris29** [reply]
It doesn't show any errors when reading the controller pak and using the controller test feature works so I believe the connection is fine for the controller adapter to the board...anything else I could check?

This is US btw.

**Comment by sanni** [comment]
I can confirm that reading/writing Controller Pak only works on HW3 right now. HW4/5 have read/write errors.

You can read and write the MPK file again with this version as temporary fix:.

**Reply by Mythris29** [reply]
Thank you for looking into this. I got the following when attempting to upload the sketch using this fix version. The sketch uploaded and it appeared to save the controller pak but then when I tried to write I got a blank screen.

C:\Users\Brian\Downloads\V9.4_Portable_SdFatV1\Arduino IDE Portable\portable\sketchbook\Cart_Reader\Cart_Reader.ino:145:11: warning: ISO C++11 requires whitespace after the macro name
 #define no-intro
           ^
C:\Users\Brian\Downloads\V9.4_Portable_SdFatV1\Arduino IDE Portable\portable\sketchbook\Cart_Reader\Cart_Reader.ino:427:10: warning: extra tokens at end of #ifdef directive
 #ifdef no-intro
          ^
C:\Users\Brian\Downloads\V9.4_Portable_SdFatV1\Arduino IDE Portable\portable\sketchbook\Cart_Reader\Cart_Reader.ino:587:10: warning: extra tokens at end of #ifdef directive
 #ifdef no-intro
          ^
C:\Users\Brian\Downloads\V9.4_Portable_SdFatV1\Arduino IDE Portable\portable\sketchbook\Cart_Reader\N64.ino: In function 'void readRom_N64()':
C:\Users\Brian\Downloads\V9.4_Portable_SdFatV1\Arduino IDE Portable\portable\sketchbook\Cart_Reader\N64.ino:3542:41: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
   if (compareCRC("n64.txt", crcStr, 1, 0)) {

**Comment by sanni** [comment]
Try the new V9.5

**Reply by Mythris29** [reply]
This version seemed to upload perfectly.

However it still seems to be hit or miss. I was able to read from card a and write to card b and testing on Mario Kart 64 the data seemed to be intact. However if I tried deleting something on card b and copying it via the cart reader I got a blank file. I also tried reading card a again and got a blank file...then got a populated file so it seems that there might be something still going on.

But it does seem like progress is being made. Just to clarify this has to be in 3V mode right?

**Comment by Mythris29** [comment]
[MPK.zip](https://github.com/sanni/cartreader/files/9366189/MPK.zip)

**Reply by Mythris29** [reply]
This is a copy of the blank and then populated file I was able to copy to a card.

**Comment by sanni** [comment]
Since it's zeros from the beginning.,maybe I put the delay in the wrong spot
Try adding another delay before readBlock here to prevent read errors:

https://github.com/sanni/cartreader/blob/master/Cart_Reader/N64.ino#L2017

````
// Controller paks, which all have 32kB of space, are mapped between 0x0000 – 0x7FFF
  for (word i = 0x0000; i < 0x8000; i += 32) {
    // Delay to prevent read errors
    delay(1);

    // Read one block of the Controller Pak into array myBlock
    readBlock(i);

    // Delay to prevent write errors
    delay(1);

    // Write block to SD card
    for (byte j = 0; j < 32; j++) {
      myFile.write(myBlock[j]);
    }
````

Or maybe the delays are not long enough, maybe change the delays from delay(1); to delay(10); 

https://github.com/sanni/cartreader/blob/master/Cart_Reader/N64.ino#L2022
https://github.com/sanni/cartreader/blob/master/Cart_Reader/N64.ino#L2088
https://github.com/sanni/cartreader/blob/master/Cart_Reader/N64.ino#L2120

**Reply by Mythris29** [reply]
Tried adding that delay and after that didn't work changing the delay to 10 on all 3 of those settings but still isn't working. In fact it's all blank files now.

**Comment by sanni** [comment]
Are you using an original Nintendo Controller and Controller Pak or some 3rd Party stuff?

**Reply by sanni** [reply]
Made some changes to the timings so the Cart Reader behaves a little bit more like a real N64.

**Reply by Mythris29** [reply]
First party on both the controllers and memory cards. 

And it looks like this new build fixed it! I did notice I had to do a full power off between reading/writing/switching cards but otherwise it seems to work really well now :) Thank you!

**Reply by sanni** [reply]
Same here, I'll add a warning to the Wiki to remind people to turn off the Cart Reader for switching Controller Paks. 👍

**Reply by Ancyker** [reply]
> I'll add a warning to the Wiki to remind people to turn off the Cart Reader

Is pressing the reset button on the display panel the same as turning it off?

**Reply by Mythris29** [reply]
> > I'll add a warning to the Wiki to remind people to turn off the Cart Reader
> 
> Is pressing the reset button on the display panel the same as turning it off?

Not in my experience testing it today. I had to actually flip the power switch each time. Pressing the reset button didn't work.

**Reply by sanni** [reply]
I added a function to reset the Controller before accessing the Controller Pak, now you don't need to power cycle.

**Reply by Mythris29** [reply]
It worked! I was able to dump then write a save file to the controller pak without having to power down the cart reader. Well done!

**Reply by Ancyker** [reply]
@sanni could you possibly add an option/function somewhere to validate that the clockgen and CIC are functioning properly? My clockgen seems to be working but I don't have any games to check if the CIC is working...

**Comment by PW5190** [comment]
Does the OSCR support the Forever Pak 64 by 4Layer Technologies? In case you don't know what that is, it's an N64 Controller Pak that features non-volatile memory (does not feature a battery) and allows save files to be kept regardless of age, hence the name "forever".
